File: Android/app/src/main/java/jp/co/riso/smartdeviceapp/SmartDeviceApp.java
/*
 * Copyright (c) 2014 RISO, Inc. All rights reserved.
 *
 * SmartDeviceApp.java
 * SmartDeviceApp
 * Created by: a-LINK Group
 */

package jp.co.riso.smartdeviceapp;

import jp.co.riso.android.util.Logger;

import android.app.Activity;
import android.app.Application;
import android.content.Context;
import android.os.Bundle;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

/**
 * @class SmartDeviceApp
 * 
 * @brief Application class.
 */
public class SmartDeviceApp extends Application implements Application.ActivityLifecycleCallbacks {

    private static volatile Context sContext;
    private static volatile Activity sActivity = null;
    
    @Override
    public void onCreate() {
        super.onCreate();
        SmartDeviceApp.sContext = getApplicationContext();
        
        if (AppConstants.DEBUG) {
            Logger.initialize(Logger.LOGLEVEL_VERBOSE, AppConstants.FOR_PERF_LOGS, AppConstants.FOR_PERF_LOGS);
            Logger.runDeleteTask(getApplicationContext());
        }
        registerActivityLifecycleCallbacks(this);
    }

    /**
     * @brief Retrieve the application context.
     * 
     * @return Application shared context
     */
    public static Context getAppContext() {
        return SmartDeviceApp.sContext;
    }

    public static Activity getActivity() { return sActivity; }

    @Override
    public void onActivityCreated(@NonNull Activity activity, @Nullable Bundle savedInstanceState) {
        sActivity = activity;
    }

    @Override
    public void onActivityStarted(@NonNull Activity activity) {
        sActivity = activity;
    }

    @Override
    public void onActivityResumed(@NonNull Activity activity) {
        sActivity = activity;
    }

    @Override
    public void onActivityPaused(@NonNull Activity activity) {}

    @Override
    public void onActivityStopped(@NonNull Activity activity) {}

    @Override
    public void onActivitySaveInstanceState(@NonNull Activity activity, @NonNull Bundle outState) {}

    @Override
    public void onActivityDestroyed(@NonNull Activity activity) {}
}

File: Android/app/src/main/java/jp/co/riso/smartdeviceapp/SmartDeviceApp.kt
/*
 * Copyright (c) 2022 RISO, Inc. All rights reserved.
 *
 * SmartDeviceApp.kt
 * SmartDeviceApp
 * Created by: a-LINK Group
 */
package jp.co.riso.smartdeviceapp

import android.app.Application
import android.app.Application.ActivityLifecycleCallbacks
import android.app.Activity
import android.os.Bundle
import android.annotation.SuppressLint
import android.content.Context
import jp.co.riso.android.util.Logger
import kotlin.jvm.Volatile

/**
 * @class SmartDeviceApp
 * 
 * @brief Application class.
 */
class SmartDeviceApp : Application(), ActivityLifecycleCallbacks {

    override fun onCreate() {
        super.onCreate()
        appContext = applicationContext
        if (AppConstants.DEBUG) {
            Logger.initialize(
                Logger.LOGLEVEL_VERBOSE,
                AppConstants.FOR_PERF_LOGS,
                AppConstants.FOR_PERF_LOGS
            )
            Logger.runDeleteTask(applicationContext)
        }
        registerActivityLifecycleCallbacks(this)
    }

    override fun onActivityCreated(activity: Activity, savedInstanceState: Bundle?) {
        Companion.activity = activity
    }

    override fun onActivityStarted(activity: Activity) {
        Companion.activity = activity
    }

    override fun onActivityResumed(activity: Activity) {
        Companion.activity = activity
    }

    override fun onActivityPaused(activity: Activity) {}
    override fun onActivityStopped(activity: Activity) {}
    override fun onActivitySaveInstanceState(activity: Activity, outState: Bundle) {}
    override fun onActivityDestroyed(activity: Activity) {}

    companion object {
        /**
         * @brief Retrieve the application context.
         *
         * @return Application shared context
         */
        @SuppressLint("StaticFieldLeak")
        @Volatile
        var appContext: Context? = null
            private set

        @SuppressLint("StaticFieldLeak")
        @Volatile
        var activity: Activity? = null
            private set
    }
}